### PR TITLE
Database Management 

### PR DESCRIPTION
**Pull request by brandonshearin**
## Description

This changeset adds a new page underneath "administration" that enables a user to obliterate their data. 

**3/1 UPDATE** 
the api contract has changed slightly to accommodate more flexibility in deleting asset group selectors.  the request body used to have a boolean field named `deleteHighValueSelectors`, which has been changed to the name `deleteAssetGroupSelectors` and it accepts a list of ints (corresponding to asset group id's)

the ui has been updated with 5 checkboxes instead of 4:
Screenshot 2024-03-01 at 10 49 25 AM

api changes: 
- added a new "WipeDB" permission to the "administrator" role 
- added a new endpoint `POST api/v2/clear-database` 
- added 3 new methods to the bloodhound db interface to delete data: **DeleteAllAssetGroupSelectorsForAssetGroups()** , **DeleteAllDataQuality()**, and **DeleteAllFileUploads**
- add swagger docs for new endpoint (this is going to get deleted after cody's PR so i'll keep track of that)

ui changes:
- new `<DatabaseManagement />` and `<ConfirmationDialog/>` components

## Motivation and Context

give a user the ability to light fire to their data

## How Has This Been Tested?

unit tests have been added to cover UI changes, and unit and integration tests have been added for the api handler

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: cmd/api/src/auth/permission.go
Comment (mistahj67) [inline cmd/api/src/auth/permission.go:102]: Style nit here but it seems like we follow a patter of lowercase for authority and then uppercase for name and i'm not sure if we wanted to make the authority generalized such as:
`db` and then for name `Wipe` perhaps?

File: cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx
Comment (urangel) [inline cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:201]: We should capitalize _data_ here to align with the casing in the other page titles.

File: cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx
Comment (urangel) [inline cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:231]: We use the MUI Alert components in other places in the app. This looks like a good place to use an <Alert/> so that these kinds of messages look consistent.

File: cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx
Comment (urangel) [inline cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:202]: It would be nice to swap this out for an MUI \<Box/> so that it applies the generated MUI css classes. It is effectively the same right now but if we ever start introducing any theming (like dark mode 👀 ) it will be easier to have MUI handle sweeping style changes rather than having to update this div manually. Wherever possible, try to use MUI components.

File: cmd/api/src/database/agi.go
Comment (urangel) [inline cmd/api/src/database/agi.go:48]: This is a nice clean up for readability 👍 What do you think about also moving the err declaration into the var block above? The same can be done for the change on line 180 if you decide to do so.

**Review by urangel — COMMENTED** [review]
Good job making both the API and UI changes on this! Also, thanks for all the tests you included. This feature will be a big quality of life improvement. Most, if not all, of the comments I left were minor consistency nits but the logic checks out. 

On a final note, I understand that we want to make it difficult to accidentally delete your data and that is why we ask for typing the phrase in but it felt a bit too cumbersome to me. If nothing else, maybe we can make the phrase matching case insensitive? I enjoyed the way legacy BloodHound would swap the confirm delete button with the cancel button in case one was spamming clicks. Just my two cents 🤷

**Comment by elikmiller** [comment]
I don't see our swagger docs updated with the new endpoint, those updates need to be included.

**Comment by elikmiller** [comment]
The ticket calls out that we should "initiate a new analysis" after deleting data - I think this piece is missing.

**Review by mistahj67 — CHANGES_REQUESTED** [review]
We discussed in DM but the permissions addition requires a migration to be added to create the permission as well as to add the permission to the admin role.

File: cmd/api/src/api/v2/database_management_integration_test.go
Comment (mistahj67) [inline cmd/api/src/api/v2/database_management_integration_test.go:71]: we should assert / check the read transaction has no error right? also might be good to verify that the data is first in the db and then the data is no longer in the db by doing a get around the delete?

File: cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx
Comment (benwaples) [inline cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:120]: Loving the separation of concerns here in this file!

nit: Could we pull this out to a named variable to help give context as to what we are checking? Its not very obvious at first and takes a couple reads to see whats going on!

Maybe something like 
```js
const noSelection = collectedGraphData, dataQualityHistory, fileIngestHistory, highValueSelectors].filter(Boolean).length === 0

if (noSelection) {
  ...
}

```

File: cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx
Comment (benwaples) [inline cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:90]: minor ux thing: IMO it would be nice to clear the error and input state on cancel as well!

File: cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx
Comment (benwaples) [inline cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:72]: We could reuse the `confirmationText ` variable here so if we need to change it in the future then we only need to change it in one place!!

File: cmd/api/src/api/registration/v2.go
Comment (elikmiller) [inline cmd/api/src/api/registration/v2.go:134]: My opinion is that this endpoint should be renamed. To me "database management" implies a lot more beyond clearing data. Maybe "/api/v2/clear-database"?

File: cmd/api/src/api/v2/database_management_integration_test.go
Comment (elikmiller) [inline cmd/api/src/api/v2/database_management_integration_test.go:1]: Should be 2024 for new files

File: cmd/api/src/api/v2/database_management.go
Comment (elikmiller) [inline cmd/api/src/api/v2/database_management.go:90]: What is the expectation if both `payload.CollectedGraphData` and `payload.HighValueSelectors` are true but an error occurs inside this block? Do we end up with only some data being deleted instead of all requested data deleted?

File: cmd/api/src/api/v2/database_management.go
Comment (elikmiller) [inline cmd/api/src/api/v2/database_management.go:139]: Flagging @juggernot325 since he is up-to-date with all things audit logging. Can you take a look here and validate that this is the appropriate way to create audit logs in this scenario?

File: cmd/api/src/auth/permission.go
Comment (elikmiller) [reply cmd/api/src/auth/permission.go:102]: I agree that the naming here could be improved

File: cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx
Comment (elikmiller) [reply cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:90]: The Cancel and Confirm buttons should have two different colors

File: cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx
Comment (elikmiller) [inline cmd/ui/src/views/DatabaseManagement/ConfirmationDialog.tsx:82]: We should clear this message once the user types the appropriate phrase. It should also re-appear if the phrase is subsequently modified and is no longer matching.

File: cmd/api/src/api/v2/database_management.go
Comment (elikmiller) [reply cmd/api/src/api/v2/database_management.go:90]: Also, it looks like the custom selector is deleted but the nodes themselves remained tagged as part of the asset group. Is that expected behavior @StephenHinck ?

**Review by elikmiller — CHANGES_REQUESTED** [review]

File: cmd/api/src/api/v2/database_management.go
Comment (juggernot325) [reply cmd/api/src/api/v2/database_management.go:139]: I know we discussed this while you were working on it and decided on generating the audit logs in this method. After looking at the PR as a whole I think we may want to go a different route. One thing that I think is big concern is that if there are any errors in the above operations there will not be an audit log generated at all, so that should be addressed.

I actually think this is going to need a novel implementation of audit logging given the complexity of this operation. Let's meet up and discuss some options.

File: cmd/api/src/database/db.go
Comment (brandonshearin) [reply cmd/api/src/database/db.go:155]: i removed this code. it was leftover from something i was trying that was testing related, but i ended up not going that direction

File: cmd/api/src/api/v2/database_management.go
Comment (brandonshearin) [reply cmd/api/src/api/v2/database_management.go]: to your first question, if a user requests to delet emultiple types of data, and gets a partial failure, maybe we should surface that to the UI?

so if i delete graph data and file ingest history, and file ingest history fails, the user would see on the UI something like "we encountered an error deleting file ingest history. please try again."

**Comment by brandonshearin** [comment]
> We discussed in DM but the permissions addition requires a migration to be added to create the permission as well as to add the permission to the admin role.

appreciate the guidance on this one.  i got this working now-- i added a new migration file titled v5.7.0.sql and added the sql to create the db.Wipe permission and then associated that permission with the Admin role

**Comment by brandonshearin** [comment]
> The ticket calls out that we should "initiate a new analysis" after deleting data - I think this piece is missing.

added this to the code block for deleting graph data and for deleting custom high value selectors

**Comment by mistahj67** [comment]
> I don't see our swagger docs updated with the new endpoint, those updates need to be included.

Might be worth coordinating with @codydbentley since he's knee deep in refactoring swagger currently

File: cmd/api/src/api/v2/database_management.go
Comment (juggernot325) [reply cmd/api/src/api/v2/database_management.go]: The latest iteration of this looks good to me! It should successfully capture the intent audit log and capture actual results.

File: cmd/api/src/api/v2/database_wipe.go
Comment (brandonshearin) [reply cmd/api/src/api/v2/database_wipe.go]: got it!

File: cmd/api/src/api/v2/database_wipe_integration_test.go
Comment (brandonshearin) [reply cmd/api/src/api/v2/database_wipe_integration_test.go:53]: i added a lot of unit test coverage to this function to cover more bases.  our integration test suite like lab doesn't support writing a test like "create some high value selectors in the db, call the wipe database endpoint, and then assert that there are no high value selectors in the db."

i tried to follow this approach, but what i was running into was a race condition in the test.  i would setup some fixtures in lab to create some data, i.e. some high value selectors, but asset group selectors depend on a FK relationship to an asset group, so I made a fixture to create the asset group first, and then a selector.

some really wonky behavior happened where the asset group record would be created in memory in gorm but wouldn't be committed to the db, so there was a race condition when trying to create the selector.  creating the selector usually threw a FK constraint violation because the asset group wasn't actually committed to the db yet.

it was too burdensome to try to get something to work for the integration test, so i just unit tested instead!

File: cmd/api/src/database/dataquality.go
Comment (brandonshearin) [reply cmd/api/src/database/dataquality.go:190]: sure can! done

File: cmd/api/src/database/dataquality.go
Comment (mistahj67) [inline cmd/api/src/database/dataquality.go:192]:
```suggestion
	return CheckError(
		s.db.WithContext(ctx).Exec("DELETE FROM ad_data_quality_aggregations; DELETE FROM ad_data_quality_stats; DELETE FROM azure_data_quality_aggregations; DELETE FROM azure_data_quality_stats;"),
	)
```

File: cmd/api/src/go.mod
Comment (mistahj67) [inline cmd/api/src/go.mod:50]: What are these go.uber.org/zap for?

File: cmd/ui/src/views/Administration/Administration.tsx
Comment (mistahj67) [inline cmd/ui/src/views/Administration/Administration.tsx:54]: Did we want to keep this as Database Management ? 🤔 Probably should check with product

**Review by mistahj67 — COMMENTED** [review]
Backend logic LGTM! Thanks for your patience with all of my feedback!

**Review by mistahj67 — APPROVED** [review]
Approving to remove my requested changes blocker but definitely get a more seasoned front end approval before merging!

File: cmd/ui/src/views/Administration/Administration.tsx
Comment (brandonshearin) [reply cmd/ui/src/views/Administration/Administration.tsx:54]: yep! this is the name specified in the acceptance criteria

File: cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx
Comment (maffkipp) [inline cmd/ui/src/views/DatabaseManagement/DatabaseManagement.tsx:1]: I think it might be helpful to break the reducer, hook, and component out into separate files here

File: cmd/api/src/database/agi.go
Comment (mistahj67) [inline cmd/api/src/database/agi.go:218]: We need the `WithContext(ctx)` here on this db call

Comment (brandonshearin) [reply cmd/api/src/database/agi.go]: gotcha

File: cmd/api/src/api/v2/database_wipe.go
Comment (elikmiller) [inline cmd/api/src/api/v2/database_wipe.go:38]: If the purpose of this endpoint is to "delete high value selectors" - why is the onus on the requestor to provide the correct asset group ID? Can the service not derive the asset group ID for the appropriate asset group?

File: cmd/api/src/api/v2/database_wipe.go
Comment (elikmiller) [inline cmd/api/src/api/v2/database_wipe.go:68]: Assuming the answer to my previous question is "the service should not need the requestor to provide an asset group ID" then we can remove this validation

**Review by elikmiller — CHANGES_REQUESTED** [review]

File: cmd/api/src/api/v2/database_wipe.go
Comment (StephenHinck) [reply cmd/api/src/api/v2/database_wipe.go:38]: I forgot to consider in the AC for this ticket that multiple groups exist (and will become more prevalent in the future).

Would it make more sense to leave it this way, change `DeleteHighValueSelectors` to `DeleteGroupSelectors` and then move `AssetGroupId` to an array/list for improved long-term supportability (with an 'all' option)? I recognize this would be a change to AC and scope, and I am happy to discuss it.

File: cmd/api/src/api/v2/database_wipe.go
Comment (brandonshearin) [reply cmd/api/src/api/v2/database_wipe.go]: removed!

File: cmd/api/src/api/v2/database_wipe.go
Comment (elikmiller) [reply cmd/api/src/api/v2/database_wipe.go]: @StephenHinck it is your feature/call to make. I think that what @brandonshearin has implemented satisfies your original AC but if that is no longer desired then we will need time to discuss and make adjustments. Consider that once this changeset is in and released it will be harder to make changes to the API contract.

**Review by elikmiller — APPROVED** [review]

**Pull request merged**
